Change build system to CMake

siggmo opened the pull request:
## Why CMake?
CMake currently is the defacto standard for building C++ projects. It abstracts away the details of different build systems (like traditional Makefiles, Ninja, Visual Studio, ...) using an own, more verbose syntax. It also provides additional features to help with common tasks such as testing, resolving dependencies, installing and packaging.

## Implemented Features
* Build project
* .gitignore file to ignore build artifacts and other files that should not be included in a version control system
* Option for a shared build
* Option to exclude examples from build
* Automatically check for dependencies and determine required paths
* Install library systemwide (on linux e.g. to /usr/local) including headers
* Other projects can consume ConcentricInterpolation using `FetchContent` or `find_package()` and then easily link to it using `target_link_librares()`
* Semantic versioning strategy with compatibility set to `SameMajorVersion` (what version number should we start with? currently I set it to 0.1.0)
* Usage of CMakePresets to simplify calls to CMake (specify e.g. build directory and flags in a .json file)

Also I rewrote the README file in markdown and made the required changes in the compilation instructions.

## Not (yet) implemented
* Packaging
* Testing
* [Git version tracking](https://github.com/andrew-hardin/cmake-git-version-tracking/tree/master)
* ...

siggmo committed:
add .gitignore

siggmo committed:
add flag -fPIC to link lib into shared libs

siggmo committed:
switch to cmake

siggmo committed:
add install functionality to cmake build system

siggmo committed:
remove PRESET_CXX_FLAGS as it's unnecessary

siggmo committed:
replace PRESET_CXX_FLAGS by CMAKE_CXX_FLAGS

siggmo committed:
only build examples when top-level project

siggmo committed:
install custom find script to find package LAPACKE

siggmo committed:
fix typo and rmeove some comments

siggmo committed:
add option for a shared build

siggmo committed:
Concretize build instructions

siggmo committed:
make README markdown formatted

siggmo committed:
minor fixes

siggmo committed:
add option to use legacy Makefile linker paths

siggmo committed:
add option to use MKL headers instead of standard lapacke and cblas headers

siggmo committed:
Remove unnecessary dependencies (as they are included in the other packages)

siggmo committed:
Apply changes from FANS and FETiger. Improve dependency checking and setup (MKL/No MKL)

siggmo committed:
add pthreads link, rename cmake config file

siggmo committed:
Add target for LAPACK::LAPACK, and add comment on Intel MKL detection issue

siggmo committed:
rename install config script

siggmo committed:
autoformat

siggmo committed:
remove presets file

siggmo committed:
Readd option to build shared

siggmo committed:
update README for update CMake recipe

siggmo committed:
fix heading level

siggmo committed:
add precommit checks

siggmo committed:
initial run of pre-commit checks

siggmo committed:
add markdown lint action

siggmo committed:
disable some markdown lint policies

siggmo committed:
add build and test github action

siggmo committed:
Add matrix options for static linkage and deps fallback

siggmo committed:
improve workflow name

siggmo committed:
Add ldd call to github action

siggmo committed:
autoformat

siggmo committed:
add cmake-format to pre-commit

siggmo committed:
remove .cmake-format.yaml bc pre-commit's cmake-format can't use yaml

siggmo committed:
change fall lib lists

OliverKunc commented:
Let me know when you are done so I can merge. Would you prefer to squash the commits or is there value in keeping your history?

siggmo commented:
I think it can be merged. I've been using it in the current state for a while now and so far no issues came up.

And I would prefer to squash and merge, the history is a bit messy and a lot of experimenting...

siggmo committed:
Be less restrictive with the Intel MKL environment setup.

Sourcing the setvars.sh script is not necessary for spack installations
of Intel MKL. There, loading the module already sets up the environment.

siggmo committed:
remove whitespaces in status message
